### ISSUE TITLE
Problems with scimax-org-babel-ipython

### ISSUE THREAD

**Issue by dangom** [jkitchin/scimax#92]
I have two issues with scimax-org-babel-ipython that I don't know how to solve. They appeared after I updated my scimax repo today.
I think the issues are all somewhat related, so I'm writing everything in a single thread.

1. On master, I cannot execute a ipython src block. An attempt at C-c C-c gives me the following error message:

> let: Invalid function: (lambda (msg &optional &rest args) (if org-babel-ipython-debug (progn (save-current-buffer (set-buffer (get-buffer-create "*ob-ipython-log*")) (org-mode) (insert (format "ob-ipython: 
> %s
> 
> " (apply (quote format) msg args)))))))
> 

It seems like the problem was introduced by commit 9d73189 .

2. Starting after commit eda22fbe (_the last good commit on my machine_) and onwards, ipython src block results are never displayed. I only get the async-running or async-queued status, but the message is never replaced by the real results. If I leave emacs on and wait for some minutes, it'll start complaining that too many files are open and I'll have to kill it.

For completion: I'm on MacOS with Emacs 26.0.5 and org-mode 9.0.8. I use Spacemacs for my configuration, where I load some of scimax functionality by calling:

```elisp
  (with-eval-after-load 'ob-ipython
    (add-to-list 'load-path "~/Elisp")
    (require 'scimax-org-babel-ipython)
    (require 'scimax-org-babel-python)
    (setq org-babel-python-command "Users/user/anaconda3/bin/pythonw3")
    (require 'ox-ipynb))
```

I'd like to say that I love this project, and if there is anything I can do on my own to help solve these problems, I'd try it. Thanks a lot!

**Comment by jkitchin** [comment]
That is pretty weird. I can't reproduce it here. Did you clone scimax into ~/Elisp?
For me, I have this setting. I don't know if that would do anything though.
org-babel-python-command = python

Do you see anything interesting if you run debug-ipython while you are waiting?

**Comment by dangom** [comment]
I clone scimax into ~/Git/scimax and have symbolic links in my ~/Elisp directory.

I see `debug-ipython` was introduced after the commit that broke things for me. 
In any case, I managed to get `ob-ipython-log` to work by changing its signature from `(ob-ipython-log (msg &optional &rest args))` to `(ob-ipython-log (msg &rest args))`, so that I can play around on master. The queue problem remains, and yes, I see interesting things from `debug-ipython`:

```elisp
Running: (*scratch* . yankee-tennis-pip-fourteen)
Queue: ((#<buffer *scratch*> . "yankee-tennis-pip-fourteen"))

* *ob-ipython-client-driver*

WARNING:tornado.access:404 POST /execute/cc76ed23-f7c1-4fcc-9a17-e42f76892e0c (::1) 1.29ms

* *ob-ipython-kernel-cc76ed23-f7c1-4fcc-9a17-e42f76892e0c*


* *Python*

Jupyter console 5.1.0

Python 3.6.1 | packaged by conda-forge | (default, May 23 2017, 14:31:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.



In [1]: python.el: native completion setup loaded

In [2]: 
* *ob-ipython-debug*

HTTP/1.1 404 Not Found
Server: TornadoServer/4.5.1
Content-Type: text/html; charset=UTF-8
Date: Thu, 08 Jun 2017 14:35:08 GMT
Content-Length: 69

<html><title>404: Not Found</title><body>404: Not Found</body></html>
* *ob-ipython-kernel-7bb52d5e-86f9-4f7f-b175-d4f24fb4b3aa*

```

And even more interestingly, `toggle-debug-on-error` gives me further info:
```elisp
Debugger entered--Lisp error: (error "There was a fatal error trying to process the request. See *ob-ipython-debug*")
  signal(error ("There was a fatal error trying to process the request. See *ob-ipython-debug*"))
  error("There was a fatal error trying to process the request. See *ob-ipython-debug*")
  ob-ipython--dump-error("HTTP/1.1 404 Not Found\nServer: TornadoServer/4.5.1\nContent-Type: text/html; charset=UTF-8\nDate: Thu, 08 Jun 2017 14:40:31 GMT\nContent-Length: 69\n\n<html><title>404: Not Found</title><body>404: Not Found</body></html>")
  (if (>= (url-http-parse-response) 400) (ob-ipython--dump-error (buffer-string)) (goto-char url-http-end-of-headers) (let* ((json-array-type (quote list)) (json (json-read))) (if (string= "error" (cdr (assoc (quote msg_type) (elt json 0)))) (progn (save-current-buffer (set-buffer (car *org-babel-async-ipython-running-cell*)) (org-babel-goto-named-src-block (cdr *org-babel-async-ipython-running-cell*)) (org-babel-remove-result)))) json))
  (ob-ipython--eval (if (>= (url-http-parse-response) 400) (ob-ipython--dump-error (buffer-string)) (goto-char url-http-end-of-headers) (let* ((json-array-type (quote list)) (json (json-read))) (if (string= "error" (cdr (assoc (quote msg_type) (elt json 0)))) (progn (save-current-buffer (set-buffer (car *org-babel-async-ipython-running-cell*)) (org-babel-goto-named-src-block (cdr *org-babel-async-ipython-running-cell*)) (org-babel-remove-result)))) json)))
  (let* ((ret (ob-ipython--eval (if (>= (url-http-parse-response) 400) (ob-ipython--dump-error (buffer-string)) (goto-char url-http-end-of-headers) (let* ((json-array-type ...) (json ...)) (if (string= "error" ...) (progn ...)) json)))) (result (cdr (assoc :result ret))) (output (cdr (assoc :output ret))) params (current-cell *org-babel-async-ipython-running-cell*) (name (cdr current-cell)) (result-type)) (save-current-buffer (set-buffer (car current-cell)) (save-excursion (org-babel-goto-named-src-block name) (setq result-type (org-babel-src-block-get-property (quote org-babel-ipython-result-type))) (org-babel-src-block-put-property (quote org-babel-ipython-executed) t) (ob-ipython-log "Got a result-type of %s\n return from the kernel:  %S" result-type ret) (setq params (car (cdr (cdr (org-babel-get-src-block-info))))) (org-babel-remove-result) (cond ((string= "output" result-type) (let ((res ...)) (if (not ...) (progn ...)))) ((string= "value" result-type) (org-babel-insert-result (cdr (assoc ... result)) (cdr (assoc :result-params ...))))) (org-redisplay-inline-images))) (setq *org-babel-async-ipython-running-cell* nil) (org-babel-async-ipython-process-queue))
  ob-ipython--async-callback((:error (error http 404)))
  apply(ob-ipython--async-callback (:error (error http 404)))
  url-http-activate-callback()
  url-http-content-length-after-change-function(146 216 70)
  url-http-wait-for-headers-change-function(1 222 221)
  url-http-generic-filter(#<process localhost> "HTTP/1.1 404 Not Found
\nServer: TornadoServer/4.5.1
\nContent-Type: text/html; charset=UTF-8
\nDate: Thu, 08 Jun 2017 14:40:31 GMT
\nContent-Length: 69
\n
\n<html><title>404: Not Found</title><body>404: Not Found</body></html>")

```

Same problem if I change `org-babel-python-command` to python. (The only reason I set it to pythonw3 is because that is conda's framework python build. The stock conda python3 doesn't play well with matplotlib's macOS backend. )

**Comment by jkitchin** [comment]
That seems to be a problem with the tornado server that driver.py sets up.

I am surprised you have trouble with the conda python. The stock python command for me runs a conda python3 executable and it works fine for all the matplotlib work I have done. Maybe I don't use a macOS backend, what do you use that for?

**Comment by dangom** [comment]
I have problems with python3 running within Emacs. Matplotlib figures do not come up to the foreground after `plt.show()`, and I have no way of seeing my plots (similar issue described [here](https://stackoverflow.com/questions/8202228/make-matplotlib-plotting-window-pop-up-as-the-active-one)). 

I learned that by changing python3 to pythonw3, as in:

```elisp
  (setq python-shell-interpreter "/Users/dangom/anaconda3/bin/pythonw3"
        python-shell-interpreter-args "-m IPython --simple-prompt -i"
        flycheck-python-pylint-executable "/Users/dangom/anaconda3/bin/pylint")
```

everything works fine.
The reason python3 doesn't work, in my poor understanding, is because only MacOS framework builds of python are capable of interacting properly with the native GUI (more info [here](https://matplotlib.org/faq/osx_framework.html)).

I don't know why on earth anaconda ships with two python3 binaries, but they are different:

```bash
$ md5 $HOME/anaconda3/bin/python3
$ md5 $HOME/anaconda3/bin/pythonw3
```

Here, only pythonw3 works as it should.

**Comment by jkitchin** [comment]
If I try to use your settings, then everything stops working for me. It seems to be the pythonw3 executable. Once I use that it stops working until I kill a bunch of emacs processes, one named localhost, one named client-driver and a kernel one.

I haven't tried making plots pop up in a while, since they are inline now.  Do you need them to pop up for interactive purposes?

**Comment by et2010** [comment]
I think this is related with https://github.com/jkitchin/scimax/issues/89, since I've also seen the same 404 error.

**Comment by dangom** [comment]
I tried with stock python3 but I still get the 404 error. =/

I don't need to make them pop up, but that is a bad habit I learned from using Matlab that I haven't eliminated completely yet. Ideally I'd like both the org-mode and the file + repl workflows to work without issues, though.

**Comment by dangom** [comment]
Wow, I don't know what happened in the last 27 days, but I've just updated scimax, and I'm surprised at how smoothly scimax-ipython is running. Thanks so much for doing this and sharing with the community.

The only thing I noticed recently is that the "%paste" magic doesn't work on the Inferior Python buffer. It does work if I call run-python, or open ipython on a terminal. The issue is not related to the latest updates though.
